File: JVM/jnetproto-java/src/main/java/jnetproto/java/BinaryReader.java
package jnetproto.java;

import jnetproto.java.compat.BitConverter;
import jnetproto.java.compat.Reflect;
import jnetproto.java.compat.Tuples;
import org.javatuples.Tuple;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Array;
import java.math.BigDecimal;
import java.nio.ByteBuffer;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.*;

public class BinaryReader implements IDataReader {
    private final Charset _enc;
    private final InputStream _stream;

    public BinaryReader(InputStream stream) {
        _enc = StandardCharsets.UTF_8;
        _stream = stream;
    }

    private byte[] readBytes(int size) throws IOException {
        var bytes = new byte[size];
        var length = _stream.read(bytes);
        if (length != size && size != 0)
            throw new IllegalArgumentException(length + " ; " + size);
        return bytes;
    }

    @Override
    public boolean readBool() throws IOException {
        return readI8() == 1;
    }

    @Override
    public byte readI8() throws IOException {
        return readBytes(1)[0];
    }

    @Override
    public short readI16() throws IOException {
        return BitConverter.toInt16(readBytes(2));
    }

    @Override
    public int readI32() throws IOException {
        return BitConverter.toInt32(readBytes(4));
    }

    @Override
    public long readI64() throws IOException {
        return BitConverter.toInt64(readBytes(8));
    }

    @Override
    public float readF32() throws IOException {
        return BitConverter.toSingle(readBytes(4));
    }

    @Override
    public double readF64() throws IOException {
        return BitConverter.toDouble(readBytes(8));
    }

    @Override
    public BigDecimal readF128() throws IOException {
        return new BigDecimal(readUtf8());
    }

    @Override
    public char readChar() throws IOException {
        return (char)readI16();
    }

    @Override
    public String readUtf8() throws IOException {
        return _enc.decode(ByteBuffer.wrap(readBytes(_stream.read()))).toString();
    }

    @Override
    public Duration readDuration() throws IOException {
        var ms = readF64();
        var cast = (long) ms;
        return Duration.ofMillis(cast);
    }

    @Override
    public LocalDateTime readTimestamp() throws IOException {
        var millis = readI64();
        var nano = readI32() * 100;
        return LocalDateTime.ofEpochSecond(millis, nano, ZoneOffset.UTC);
    }

    @Override
    public UUID readGuid() throws IOException {
        return BitConverter.toGuid(readBytes(16));
    }

    @Override
    public Object readArray() throws IOException {
        var item = Reflect.toDataType(_stream.read());
        var rank = _stream.read();
        var lengths = new int[rank];
        for (var i = 0; i < rank; i++)
            lengths[i] = readI32();
        var clazz = DataTypes.getClass(item);
        var array = Array.newInstance(clazz, lengths);
        var indices = new int[rank];
        for (var i = 0; i < rank; i++)
            for (var j = 0; j < lengths[i]; j++) {
                var obj = readObject(item);
                indices[i] = j;
                Array.set(array, indices[0], obj);
            }
        return array;
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    @Override
    public Map readMap() throws IOException {
        var keyKind = Reflect.toDataType(_stream.read());
        var valKind = Reflect.toDataType(_stream.read());
        var size = readI32();
        var map = new TreeMap();
        for (var i = 0; i < size; i++) {
            var key = readObject(keyKind);
            var val = readObject(valKind);
            map.put(key, val);
        }
        return map;
    }

    @Override
    public Tuple readTuple() throws IOException {
        var size = readI8();
        var args = new Object[size];
        for (var i = 0; i < size; i++)
        {
            var obj = readObject();
            args[i] = obj;
        }
        var method = Arrays.stream(Tuples.class.getMethods())
                .filter(m -> m.getName().equals("create") && m.getParameterCount() == args.length);
        return (Tuple)Reflect.invoke(method.findFirst().orElseThrow(), null, args);
    }

    @Override
    public Object[] readBag() throws IOException {
        var size = readI8();
        var args = new Object[size];
        for (var i = 0; i < size; i++)
        {
            var obj = readObject();
            args[i] = obj;
        }
        return args;
    }

    @Override
    public byte[] readBinary() throws IOException {
        var size = readI32();
        return readBytes(size);
    }

    @Override
    public Set readSet() throws IOException {
        return (Set)readIterable(new TreeSet());
    }

    @Override
    public List readList() throws IOException {
        return (List)readIterable(new ArrayList());
    }

    private Iterable readIterable(Iterable coll) throws IOException
    {
        var valKind = Reflect.toDataType(_stream.read());
        var size = readI32();
        var adder = Reflect.getMethod(coll, "add", Object.class);
        for (var i = 0; i < size; i++)
        {
            var val = readObject(valKind);
            Reflect.invoke(adder, coll, new Object[] { val });
        }
        return coll;
    }

    @Override
    public Object readObject() throws IOException {
        var kind = Reflect.toDataType(_stream.read());
        return readObject(kind);
    }

    private Object readObject(DataType kind) throws IOException {
        switch (kind)
        {
            case Bool: return readBool();
            case I8: return readI8();
            case I16: return readI16();
            case I32: return readI32();
            case I64: return readI64();
            case F32: return readF32();
            case F64: return readF64();
            case F128: return readF128();
            case Char: return readChar();
            case UTF8: return readUtf8();
            case Duration: return readDuration();
            case Timestamp: return readTimestamp();
            case Guid: return readGuid();
            case Array: return readArray();
            case Map: return readMap();
            case Tuple: return readTuple();
            case Set: return readSet();
            case List: return readList();
            case Bag: return readBag();
            case Binary: return readBinary();
            default: throw new IllegalArgumentException(kind.toString());
        }
    }
    
    @Override
    public void close() throws Exception {
        if (_stream == null)
            return;
        _stream.close();
    }
}

File: NET/JNetProto.Sharp/Core/BinaryReader.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using JNetProto.Sharp.API;

namespace JNetProto.Sharp.Core
{
    public class BinaryReader : IDataReader
    {
        private readonly CultureInfo _cult;
        private readonly Encoding _enc;
        private readonly Stream _stream;

        public BinaryReader(Stream stream)
        {
            _cult = CultureInfo.InvariantCulture;
            _enc = Encoding.UTF8;
            _stream = stream;
        }

        private byte[] ReadBytes(int size)
        {
            var bytes = new byte[size];
            var length = _stream.Read(bytes);
            if (length != size)
            {
                var hex = bytes.Length <= 1024
                    ? BitConverter.ToString(bytes).Replace("-", "")
                    : "(too big for hex)";
                throw new ArgumentException($"Got {length} B instead of {size}! {hex}");
            }
            return bytes;
        }

        public bool ReadBool()
        {
            return ReadI8() == 1;
        }

        public byte ReadI8()
        {
            return ReadBytes(1)[0];
        }

        public short ReadI16()
        {
            return BitConverter.ToInt16(ReadBytes(2));
        }

        public int ReadI32()
        {
            return BitConverter.ToInt32(ReadBytes(4));
        }

        public long ReadI64()
        {
            return BitConverter.ToInt64(ReadBytes(8));
        }

        public float ReadF32()
        {
            return BitConverter.ToSingle(ReadBytes(4));
        }

        public double ReadF64()
        {
            return BitConverter.ToDouble(ReadBytes(8));
        }

        public decimal ReadF128()
        {
            return decimal.Parse(ReadUtf8(wide: false), _cult);
        }

        public char ReadChar()
        {
            return (char)ReadI16();
        }

        public string ReadUtf8()
        {
            return ReadUtf8(wide: true);
        }

        private string ReadUtf8(bool wide)
        {
            var size = wide ? ReadI16() : ReadI8();
            return _enc.GetString(ReadBytes(size));
        }

        public TimeSpan ReadDuration()
        {
            return TimeSpan.FromMilliseconds(ReadF64());
        }

        public DateTime ReadTimestamp()
        {
            var date = DateTimeOffset.FromUnixTimeSeconds(ReadI64()).UtcDateTime;
            return date.AddTicks(ReadI32());
        }

        public Guid ReadGuid()
        {
            return new Guid(ReadBytes(16));
        }

        public Array ReadArray()
        {
            var item = (DataType)_stream.ReadByte();
            var rank = _stream.ReadByte();
            var lengths = new int[rank];
            for (var i = 0; i < rank; i++)
                lengths[i] = ReadI32();
            var clazz = DataTypes.GetClass(item);
            var array = Array.CreateInstance(clazz, lengths);
            var indices = new int[rank];
            for (var i = 0; i < rank; i++)
            for (var j = 0; j < lengths[i]; j++)
            {
                var obj = ReadObject(item);
                indices[i] = j;
                array.SetValue(obj, indices);
            }
            return array;
        }

        public IDictionary ReadMap()
        {
            var keyKind = (DataType)_stream.ReadByte();
            var valKind = (DataType)_stream.ReadByte();
            var size = ReadI32();
            var keyClass = DataTypes.GetClass(keyKind);
            var valClass = DataTypes.GetClass(valKind);
            var mapClass = typeof(SortedDictionary<,>).MakeGenericType(keyClass, valClass);
            var map = (IDictionary)Activator.CreateInstance(mapClass)!;
            for (var i = 0; i < size; i++)
            {
                var key = ReadObject(keyKind);
                var val = ReadObject(valKind);
                map[key] = val;
            }
            return map;
        }

        public ITuple ReadTuple()
        {
            var size = ReadI8();
            var args = new object[size];
            var types = new Type[size];
            for (var i = 0; i < size; i++)
            {
                var obj = ReadObject();
                args[i] = obj;
                types[i] = obj.GetType();
            }
            var method = typeof(Tuple).GetMethods()
                .First(m => m.Name == nameof(Tuple.Create) && m.GetParameters().Length == types.Length);
            var tuple = (ITuple)method.MakeGenericMethod(types).Invoke(null, args);
            return tuple;
        }

        public object[] ReadBag()
        {
            var size = ReadI8();
            var args = new object[size];
            for (var i = 0; i < size; i++)
            {
                var obj = ReadObject();
                args[i] = obj;
            }
            return args;
        }

        public byte[] ReadBinary()
        {
            var size = ReadI32();
            return ReadBytes(size);
        }

        public object ReadNull()
        {
            return null;
        }

        public IEnumerable ReadSet()
        {
            var setType = typeof(SortedSet<>);
            return ReadEnumerable(setType);
        }

        public IList ReadList()
        {
            var listType = typeof(List<>);
            return (IList)ReadEnumerable(listType);
        }

        private IEnumerable ReadEnumerable(Type type)
        {
            var valKind = (DataType)_stream.ReadByte();
            var size = ReadI32();
            var valClass = DataTypes.GetClass(valKind);
            var genType = type.MakeGenericType(valClass);
            var coll = (IEnumerable)Activator.CreateInstance(genType)!;
            var adder = genType.GetMethod("Add", new[] { valClass })!;
            for (var i = 0; i < size; i++)
            {
                var val = ReadObject(valKind);
                adder.Invoke(coll, new[] { val });
            }
            return coll;
        }

        public object ReadObject()
        {
            var kind = (DataType)_stream.ReadByte();
            return ReadObject(kind);
        }

        private object ReadObject(DataType kind)
        {
            switch (kind)
            {
                case DataType.Bool: return ReadBool();
                case DataType.I8: return ReadI8();
                case DataType.I16: return ReadI16();
                case DataType.I32: return ReadI32();
                case DataType.I64: return ReadI64();
                case DataType.F32: return ReadF32();
                case DataType.F64: return ReadF64();
                case DataType.F128: return ReadF128();
                case DataType.Char: return ReadChar();
                case DataType.UTF8: return ReadUtf8();
                case DataType.Duration: return ReadDuration();
                case DataType.Timestamp: return ReadTimestamp();
                case DataType.Guid: return ReadGuid();
                case DataType.Array: return ReadArray();
                case DataType.Map: return ReadMap();
                case DataType.Tuple: return ReadTuple();
                case DataType.Set: return ReadSet();
                case DataType.List: return ReadList();
                case DataType.Bag: return ReadBag();
                case DataType.Binary: return ReadBinary();
                case DataType.Null: return ReadNull();
                default: throw new ArgumentException($"{nameof(ReadObject)} {kind}");
            }
        }

        public void Dispose()
        {
            _stream?.Dispose();
        }
    }
}
